_posts/2024-10-23-whats-new-in-the-vss-extension.md

```diff
@@ -0,0 +1,233 @@
+                             └─────────────┬─────────────┘
+                             ┌─────────────┴─────────────┐
+                             │       CROSS_PRODUCT       ├──────────────┐
+                             └─────────────┬─────────────┘              │
+                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
+                             │         SEQ_SCAN          ││         DELIM_SCAN        │
+                             │    ────────────────────   ││    ────────────────────   │
+                             │        ~10000 Rows        ││         ~5000 Rows        │
+                             └───────────────────────────┘└───────────────────────────┘
+```
+</details>
+
+While this plan looks very complicated, the most worrysome among these operators is the `CROSS_PRODUCT` towards the bottom of the plan, which blows up the expected cardinality and is a sign that we are doing a lot of work that we probably don't want to do. However, if we create an HNSW index on the `items` table using
+
+```sql
+CREATE INDEX my_hnsw_idx ON items USING HNSW(embedding);
+```
+
+and re-run `EXPLAIN`, we get this plan instead:
+
+<details markdown='1'>
+<summary markdown='span'>
+Query plan with HNSW index (operators and expected cardinalities)
+</summary>
+
+```text
+┌───────────────────────────┐
+│         PROJECTION        │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│       HASH_GROUP_BY       │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│      HNSW_INDEX_JOIN      │
+│    ────────────────────   │
+│        ~50000 Rows        │
+└─────────────┬─────────────┘
+┌─────────────┴─────────────┐
+│         SEQ_SCAN          │
+│    ────────────────────   │
+│        ~10000 Rows        │
+└───────────────────────────┘
+```
+</details>
+
+We can see that this plan is drastically simplified, but most importantly, the new `HNSW_INDEX_JOIN` operator replaces the `CROSS_PRODUCT` node that was there before and the estimated cardinality went from 5,000,000 to 50,000! Executing this query now takes about 0.15 seconds. That's an almost 66× speedup!
+
+This optimization was just recently added to the VSS extension, so if you've already installed `vss` for DuckDB v1.1.2, run the following command to get the latest version:
+
+```sql
+FORCE INSTALL vss;
+```
```

Comment:
I would go with `UPDATE EXTENSIONS` here, that is the better version of this (even tough it has side effects, those are mostly positive).

UPDATE EXTENSIONS updates all currently installed extensions but only if they are changed (check is an unexpensive GET with Etag).

UPDATE EXTENSIONS basically pays only if stuff has changed, so repeated usage have ammortized low cost, while FORCE INSTALL will still download the extension in every case.